### PR TITLE
Processing data in models

### PR DESCRIPTION
**Pull request by Jolak5**
In this PR, I: 

- Create model classes for all entities as shown in the [ERD diagram]
- Set up associations between models.

### CODE REVIEW

**Review by MaganaAsiati — CHANGES_REQUESTED** [review]
# STATUS: Required Changes ♻️♻️♻️ 

Hi @Jolak5 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

AlonzoLeroneThisCloseGIF

## Highlights
✅ No linter error
✅ README is professional

## Required Changes ♻️

- Kindly implement all the changes I list below to make your project completed, you did a great job by adding the line to implement functionality however make sure you add them as methods each of them must be wrapped in def...end, 

- Kindly make sure that your User model has a method that returns the 3 most recent posts.  :+1: 

```ruby
def recent_post
  posts.order('created_at Desc').limit(3)
end
```

- Kindly make sure that the post model has a method that updates the posts counter for a user.

```ruby
def update_post_counter
  author.increment!(:posts_counter)
end
```
- Kindly make sure that the post model has a method that returns the 5 most recent comments for a given post.

   ```ruby
   
   ...
   
   def recent_comments
     comments.order('created_at Desc').limit(5)
   end
   ```
- Kindly make sure that the comment model has a method that updates the comments counter for a post.

   ```ruby

   ...
   
   def update_comments_counter
     post.increment!(:comments_counter)
   end
   ```



- kindly make sure that the post model has a method that returns the 5 most recent comments for a given post. :+1: 

   ```ruby
 
   ...
   
   def recent_comments
     comments.order('created_at Desc').limit(5)
   end
   ```
- Kindly make sure that the comment model has a method that updates the comments counter for a post. :+1: 

   ```ruby

   ...
   
   def update_comments_counter
     post.increment!(:comments_counter)
   end
   ```

  
- Kindly make sure that the Like model has a method that updates the likes counter for a post. :+1: 

   ```ruby
 
   ...
   
   def update_likes_counter
     post.increment!(:likes_counter)
   end
   ```


_Check the comments under the review._

## Optional suggestions

-  It would be nice to  use one of the [Rails Active Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html) -* `after_save` to call the `#update_post_counter` method every time a new post has been saved to the database, then make the `update_post_counter` method private

   ```ruby
     # app/models/post.rb
     after_save :update_post_counter
   
     private
   
     def update_post_counter
       author.increment!(:posts_counter)
     end
   ```

- Kindly make sure in the future you make all the changes need to implement in the current branch you submit for review so that it will be easy for the reviewer to make inline comments and see directly the changes you did :+1: 
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
```

File: app/models/user.rb [hunk 1]
```diff
@@ -3,5 +3,13 @@ class User < ApplicationRecord
   has_many :comments, dependent: :destroy, foreign_key: :author_id
   has_many :likes, dependent: :destroy, foreign_key: :author_id
 
-  scope :user_recent_posts, -> { find(1).posts.order(created_at: :desc).first(3) }
+  # scope :user_recent_posts, -> { find(1).posts.order(created_at: :desc).first(1) }
+  scope :user_recent_posts, -> { find(1).posts.order(created_at: :desc).limit(3) }
```
Comment (TedLivist) [inline app/models/user.rb]: - [ ] Kindly remove this comment and this scope (which is incorrect anyways). You already have `recent_post` method. Let's stick to that.

File: app/models/post.rb [hunk 2]
```diff
@@ -3,6 +3,23 @@ class Post < ApplicationRecord
   has_many :posts
   has_many :comments
 
-  scope :update_post_counter, -> { User.find(2).update(posts_counter: User.find(2).posts.count) }
-  scope :post_recent_comments, -> { Comment.where(post_id: 2).order(created_at: :desc).first(5) }
+  scope :update_posts_counter, -> { User.find(2).update(postsCounter: User.find(2).posts.count) }
+  scope :post_recent_comments, -> { Comment.where(post_id: 1).order(created_at: :desc).limit(5) }
```
Comment (TedLivist) [inline app/models/post.rb]: - [ ] Let's remove these scopes. We already have `recent_comments` and `update_post_counter` methods that do what we intended for these queries.

File: app/models/like.rb [hunk 3]
```diff
@@ -2,5 +2,9 @@ class Like < ApplicationRecord
   belongs_to :author, class_name: 'User', foreign_key: :author_id
   belongs_to :posts
 
+  def update_likes_counter
+    post.increment!(:likesCounter)
+  end
+
   scope :update_post_likes, -> { Post.find(2).update(likesCounter: Like.where(post_id: 2).size) }
```
Comment (TedLivist) [inline app/models/like.rb]: - [ ] Kindly remove this scope as there is custom method that takes care of this query.

File: app/models/like.rb [hunk 4]
```diff
@@ -2,5 +2,9 @@ class Like < ApplicationRecord
   belongs_to :author, class_name: 'User', foreign_key: :author_id
   belongs_to :posts
 
+  def update_likes_counter
+    post.increment!(:likesCounter)
+  end
```
Comment (TedLivist) [inline app/models/like.rb]: - [ ] **[OPTIONAL]** As the previous reviewer suggested, I would recommend you add an callback for this method

  ```
  after_save :update_likes_counter  
  
  private

  def update_likes_counter
    ...
  end
  ```

File: app/models/comment.rb [hunk 5]
```diff
@@ -2,5 +2,9 @@ class Comment < ApplicationRecord
   belongs_to :author, class_name: 'User', foreign_key: :author_id
   belongs_to :post
 
+  def update_comments_counter
+    post.increment!(:commentsCounter)
+  end
+
   scope :update_comment_counter, -> { Post.find(3).update(commentsCounter: Comment.where(post_id: 3).size) }
```
Comment (TedLivist) [inline app/models/comment.rb]: - [ ] Kindly remove this scope. There is no need having duplicate methods that do the same thing.

File: app/models/comment.rb [hunk 6]
```diff
@@ -2,5 +2,9 @@ class Comment < ApplicationRecord
   belongs_to :author, class_name: 'User', foreign_key: :author_id
   belongs_to :post
 
+  def update_comments_counter
+    post.increment!(:commentsCounter)
+  end
```
Comment (TedLivist) [inline app/models/comment.rb:7]: - [ ] **[OPTIONAL]** I would recommend you add a callback for this method based on previous suggestions.

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @Jolak5,

✔️ Establishment of associations between models

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the **[OPTIONAL]** prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by TedLivist — APPROVED** [review]
## STATUS: APPROVED ✔️ ✔️

Hi @Jolak5,

✔️ Use of custom methods

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**